npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,442 @@
+        port: 443
+---
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: BaselineAdminNetworkPolicy
+metadata:
+  name: default
+spec:
+  subject:
+    namespaces: {}
+  ingress:
+    - action: Deny
+      to:
+      - networks:
+        - "0.0.0.0/0"
+```
+
+### Expected Behavior
+
+1. A FQDN egress policy does not grant the workload permission to communicate
+   with any in-cluster DNS services (like `kube-dns`). A separate rule needs to
+   be configured to allow traffic to any DNS servers.
+1. FQDN policies should not affect the ability of workloads to resolve domains,
+   only their ability to communicate with the IP backing them.
+   *  For example, if a policy allows traffic to `kubernetes.io`, any selected
+      Pods can still resolve `wikipedia.org` or
+      `my-services.default.svc.cluster.local`, but can not send traffic to them
+      unless allowed by a different rule.
+1. Each implementation will provide guidance on which DNS name-server is
+   considered authoritative for resolving domain names. This could be the
+   `kube-dns` Service or potentially some other DNS provider specified in the
+   implementation's configuration.
+1. DNS record querying and lifetimes:
+   *  Pods are expected to make a DNS query for a domain before sending traffic
+      to it. If the Pod fails to send a DNS request and instead just sends
+      traffic to the IP (either because of caching or a static config), traffic
+      is not guaranteed to flow.
+   *  Pods should respect the TTL of DNS records they receive. Trying to
+      establish new connection using DNS records that are expired is not
+      guaranteed to work.
+   *  When the TTL for a DNS record expires, the implementor should stop
+      allowing new connections to that IP. Existing connection will still be
+      allowed (that's consistent with NetworkPolicy behavior on long-running
+      connections). 
+1. PTR records are not required to properly configure a FQDN selector. For
+   example, as long as an A record exists mapping `my-hostname` to `1.2.3.4`,
+   the Network Policy implementation should allow traffic to `1.2.3.4`. There is
+   no requirement that a PTR record for `1.2.3.4.in-addr.arpa` exist or that it
+   points to `my-hostname` (it is allowed to point to `other-host`).
+1. Targeting in-cluster endpoints with FQDN selector:
+   *  ✅︎ Supported:
+      *  Selecting Pods using their [generated DNS
+         record](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods)
+         (for example `pod-ip-address.my-namespace.pod.cluster.local`). This is
+         analogous to selecting the Pod by its IP address using the Network
+         selector.
+      * Headless Services can be selected using their [generated DNS
+        record](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#services)
+        because the generated DNS records contain a list of all the Pod IPs that
+        back the service.
+   *  ❌ Not Supported:
```

Comment:
Just to add another gotcha: we had a user who used DNS load balancing to load balance to a set of external IPs, each in a separate k8s cluster.  They then tried to use DNS policy with _that_; it would work when a pod connected to another cluster but not the local cluster because kube-proxy was short-circuiting the load balancer and handling the external IP in-cluster.  

User thought they were using an external load balancer but that load balancer used external IPs under the covers. Very confusing!

Reply:
I love these stories :D
so did the problem happen because DNS matching happened after service loadbalancing? IOW, would matching network policy before service resolution solve this problem?

Reply:
Perhaps, but then it'd be very hard to interleave normal and DNS policy in anything that resembles iptables.  The DNS policy would happen "first" in order to see the pre-DNAT IPs but what if a higher priority normal policy should have dropped that traffic? BPF dataplanes that do policy and NAT can keep the pre-DNAT information around to match on at the policy stage.